File: problem2-bfs.py
"""
# Definition for Employee.
class Employee:
    def __init__(self, id: int, importance: int, subordinates: List[int]):
        self.id = id
        self.importance = importance
        self.subordinates = subordinates
"""
from collections import deque
class Solution:
    def getImportance(self, employees: List['Employee'], id: int) -> int:
        if not employees: return 0
        hashmap = {}
        for emp in employees:
            hashmap[emp.id] = emp
        
        q = deque()
        q.append(id)
        
        total_importance = 0
        
        while q:
            size = len(q)
            for i in range(size):
                eid = q.popleft()
                e = hashmap[eid]
                total_importance += e.importance
                for sub in e.subordinates:
                    q.append(sub)
        return total_importance
            
#Time complexity is O(n) and space complexity is O(n)
#BFS approach used for evaluating all the subordiantes of the given id.
            

File: Problem2_BFS.java
//Time Complexity: O(n), where n is the total no. of employees, max possible.
//Space Complexity: O(n)
//Code run successfully on LeetCode.

public class Problem2_BFS {

    public int getImportance(List<Employee> employees, int id) {
        
        if(employees == null || employees.size() == 0)
            return -1;
        
        HashMap<Integer, Employee> map = new HashMap<>();
        int total = 0;
        
        for(Employee e : employees){
            map.put(e.id, e);
        }
        
        Queue<Integer> q = new LinkedList<>();
        q.add(id);
        
        while(!q.isEmpty())
        {
            int curr = q.poll();    
            Employee e = map.get(curr);
            total += e.importance;
            for(int i : e.subordinates){
                q.add(i);
            }
        }
        return total;
    }
}
